### PR TITLE
Improve performance of showing partial Pipelines and runs

### LINKED ISSUE #1248: [Bug]: Improve performance of showing partial Pipelines / associated Pipeline Runs
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

When we show "the first 5" Pipelines or Pipeline Runs under a pipeline -- we don't actually filter this by network. We fetch as much as the API will allow us and just try to client side sort and manage.

### Expected Behavior

This is roughly what we need to do:
- Filter the call to the "max count" -- which is 5 today, but we should use the same constant value throughout
- Sort the response on created -- we want the latest created Pipeline or Pipeline Run -- we have other pages for them to investigate the more items that are older

This should impact these areas:
- Project Pipelines
- Project Pipelines' Runs
- Global Pipelines' Runs

### Steps To Reproduce

Setup:
1. Create 6 Pipelines
2. Inside a Pipeline create 6 Runs

View the UI in those areas and make sure your 6th one is gone and the "view more" link is available.

### Workaround (if any)

_No response_

### What browsers are you seeing the problem on?

_No response_

### Open Data Hub Version

```yml
Dashboard: v2.10.0
```

### Anything else

Somewhat related to https://github.com/opendatahub-io/odh-dashboard/issues/1223

### PR DESCRIPTION
**Pull request by uidoyen**
Closes #1248

## Description
Improve performance of showing partial Pipelines / associated Pipeline Runs by enabling pagination and limit the items. Also sort the pipelines and runs by `created_at`.

### Project Pipelines & Project Pipelines' Runs

- Created few sample pipelines and runs 
- Go to DS Project  -> select project -> Pipelines -> expand pipeline which contains the runs

Screenshot 2023-07-10 at 8 42 49 PM

Global Pipelines:

Screenshot 2023-07-10 at 11 16 59 PM

Global Pipeline **scheduled** runs:

Screenshot 2023-07-10 at 11 17 58 PM

Global Pipeline **triggered** runs(zoomed out):
Screenshot 2023-07-10 at 11 19 10 PM

## Request review criteria:

- [x] The commits have meaningful messages (squashes happen on merge by the bot).
- [x] Included any necessary screenshots or gifs if it was a UI change.
- [ ] Included tags to the UX team if it was a UI/UX change.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work
- [ ] The developer has added tests or explained why testing cannot be added (unit tests & storybook for related changes)

### CODE REVIEW

File: frontend/src/concepts/pipelines/content/tables/pipeline/PipelinesTable.tsx [hunk 1]
```diff
@@ -6,7 +6,7 @@ import PipelinesTableRow from '~/concepts/pipelines/content/tables/pipeline/Pipe
 import { FetchStateRefreshPromise } from '~/utilities/useFetchState';
 import { pipelineColumns } from '~/concepts/pipelines/content/tables/columns';
 import DeletePipelineCoreResourceModal from '~/concepts/pipelines/content/DeletePipelineCoreResourceModal';
-
+import { sortByCreatedAt } from '~/utilities/sortByCreatedAt';
```
Comment (lucferbux) [inline frontend/src/concepts/pipelines/content/tables/pipeline/PipelinesTable.tsx:9]: nit: Leave an empty space between imports and declaration

File: frontend/src/utilities/sortByCreatedAt.ts [hunk 2]
```diff
@@ -0,0 +1,2 @@
+export const sortByCreatedAt = (pipeline) =>
```
Comment (lucferbux) [inline frontend/src/utilities/sortByCreatedAt.ts:1]: pipeline is missing the type

File: frontend/src/utilities/sortByCreatedAt.ts [hunk 3]
```diff
@@ -0,0 +1,2 @@
...
+  pipeline.sort((a, b) => new Date(b.created_at).getTime() - new Date(a.created_at).getTime());
```
Comment (lucferbux) [inline frontend/src/utilities/sortByCreatedAt.ts:2]: If this method is targeted only for pipelines, add it in the `utils` file within pipelines, if not, make this method generic.

**Review by lucferbux — CHANGES_REQUESTED** [review]
I think you are missing the point of the change.
The issue describes that we need to filter the pipelines k8s call but I think part of the goal of this issue is to improve performance by not fetching all the calls right?

**Comment by lucferbux** [comment]
@uidoyen Can you rebase to get latest changes, there's a bug fix for an issue present in this PR (was in our codebase before)

File: frontend/src/concepts/pipelines/content/tables/pipeline/PipelinesTableRow.tsx [hunk 4]
```diff
@@ -32,7 +32,7 @@ const PipelinesTableRow: React.FC<PipelinesTableRowProps> = ({
 }) => {
   const navigate = useNavigate();
   const { namespace } = usePipelinesAPI();
-  const runsFetchState = usePipelineRunsForPipeline(pipeline);
+  const runsFetchState = usePipelineRunsForPipeline(pipeline, 6);
```
Comment (andrewballantyne) [inline frontend/src/concepts/pipelines/content/tables/pipeline/PipelinesTableRow.tsx:35]: No magic numbers

File: frontend/src/concepts/pipelines/content/tables/pipelineRun/PipelineRunTable.tsx [hunk 5]
```diff
@@ -32,6 +32,7 @@ const PipelineRunTable: React.FC<PipelineRunTableProps> = ({ runs }) => {
         data={filteredRuns}
         columns={pipelineRunColumns}
         enablePagination
+        minPageSize={5}
```
Comment (andrewballantyne) [inline frontend/src/concepts/pipelines/content/tables/pipelineRun/PipelineRunTable.tsx:35]: Don't do this -- data should be limited if we need to truncate the data... don't force the tables to be smaller.

**Review by andrewballantyne — CHANGES_REQUESTED** [review]
There are three views here:

* Projects Pipelines
* Global Pipelines
* Global Runs (scheduled and triggered)

There are two types of changes needed:

* Project Pipelines needs to limit to "5" because we fetching 20 has no value since we only want to show 5
* Project Pipelines & Global Pipelines show the last 5 runs inside the expanded section -- but fetch 20 and truncate client side

So we need to handle this truncation via **API** on the Project Pipelines for Pipelines & Runs. Global Pipelines we need to truncate via **API** for the expanded row's runs. Global Runs does not need to be changed for this issue.

We want to improve our fetching so we can get the "latest" from the API without having to fetch 15 more than we need. We can fetch 6 instead of 5 that we show if that's the easiest way to show the related "show more" links that take you to the Global Pipelines or Global Runs page (depending on what link you click on). 

This effort is to just limit fetching in _some_ cases for _some_ resources. We want our Global pages to show as much as it can for that given type. 

Hope this all makes sense, let me know if you have any questions.

File: frontend/src/concepts/pipelines/content/tables/pipeline/PipelinesTableExpandedRow.tsx [hunk 6]
```diff
@@ -13,6 +13,7 @@ import { PipelineKF, PipelineRunKF } from '~/concepts/pipelines/kfTypes';
 import IndentSection from '~/pages/projects/components/IndentSection';
 import { FetchState } from '~/utilities/useFetchState';
 import { usePipelinesAPI } from '~/concepts/pipelines/context';
+import { TABLE_CONTENT_LIMIT } from '~/const';
```
Comment (andrewballantyne) [inline frontend/src/concepts/pipelines/content/tables/pipeline/PipelinesTableExpandedRow.tsx:16]: This is Pipeline specific, no need to make this global. We should keep things isolated as much as we can. 

Perhaps use `~/concepts/pipelines` instead?

File: frontend/src/const.ts [hunk 7]
```diff
@@ -4,3 +4,5 @@ export const LABEL_SELECTOR_DASHBOARD_RESOURCE = `${KnownLabels.DASHBOARD_RESOUR
 export const LABEL_SELECTOR_MODEL_SERVING_PROJECT = `${KnownLabels.MODEL_SERVING_PROJECT}=true`;
 export const LABEL_SELECTOR_DATA_CONNECTION_AWS = `${KnownLabels.DATA_CONNECTION_AWS}=true`;
 export const LABEL_SELECTOR_PROJECT_SHARING = `${KnownLabels.PROJECT_SHARING}=true`;
+export const TABLE_CONTENT_LIMIT = 5;
+export const LIMIT_MAX_ITEM_COUNT = 6;
```
Comment (andrewballantyne) [inline frontend/src/const.ts:8]:
```suggestion
// Get one more than the limit to help know if there are more than we want to show
export const LIMIT_MAX_ITEM_COUNT = TABLE_CONTENT_LIMIT + 1;
```

**Comment by andrewballantyne** [comment]
@uidoyen Can you fix up your screenshots in your description?

**Review by andrewballantyne — CHANGES_REQUESTED** [review]
Couple notes

**Review by andrewballantyne — COMMENTED** [review]
Looks good code-wise and screenshot wise -- I have not tested it. 

/lgtm

**Comment by uidoyen** [comment]
@pnaik1 @manaswinidas Can you guys test it please?

**Comment by pnaik1** [comment]
Okay @uidoyen

**Comment by lucferbux** [comment]
Im gonna trigger again the checks and review it. Let's see if this time it doesn't fail

**Comment by lucferbux** [comment]
Code looks great, requests have the query params that we wanted, I only have a nit, but I think is fine as it is right now.

Screenshot 2023-07-11 at 17 53 10


As you can see in the response, the request return the total runs that you have, so `LIMIT_MAX_ITEM_COUNT` is not needed to show expanded, it can be achieved just fetching 5 and then checking if `total_size` is greater than `TABLE_CONTENT_LIMIT`.

I can imagine we haven't implemented this since `total_size` is optional and it might be missing in some scenarios right?

**Comment by lucferbux** [comment]
But im gonna approve it anyways cause code looks great.

**Review by lucferbux — APPROVED** [review]
/lgtm

**Comment by lucferbux** [comment]
I'm sorry @uidoyen I think you need to rebase interactively your PR to squash all the commits beforehand, we need to do this for ever pr for now own, please do that before merging

**Review by manaswinidas — APPROVED** [review]
The changes work fine for me.

/lgtm

**Review by andrewballantyne — APPROVED** [review]

**Pull request merged**